src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs

```diff
@@ -2531,6 +2531,27 @@ private BetterResult BetterConversionFromExpression(BoundExpression node, TypeSy
             if (!conv2.IsConditionalExpression && conv1.IsConditionalExpression)
                 return BetterResult.Right;
 
+            // C1 and C2 are collection expression conversions and the following hold:
+            // - T1 is a ref struct type with iteration type E1, and T2 is a non- ref struct type with iteration type E2, and
+            // - E1 is implicitly convertible to E2
+            if (conv1.Kind == ConversionKind.CollectionExpression &&
+                conv2.Kind == ConversionKind.CollectionExpression &&
+                _binder.TryGetCollectionIterationType((Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax)node.Syntax, t1, out TypeWithAnnotations e1) &&
```

Comment:
Is there nowhere we can retrive this information from? I'm a tad worried about the perf implications of calling this repeatedly with various overloads.

Reply:
> That would allow interpreting a [CollectionBuilder] attribute on a collection type independent of where the type is used.

I think we should seriously consider this. I'm struggling to come up with a scenario where you'd apply `CollectionBuilder` to a type and _not_ have a `GetEnumerator` on that type already, and the ability to interpret it independently of the use site is a valuable one. I'm really quite concerned about the perf of this change, and that rule change would do quite a bit to make the behavior of overload resolution predictable and explainable.

Reply:
https://github.com/dotnet/roslyn/issues/69676